zetaclient/context/chain.go

```diff
@@ -0,0 +1,177 @@
+package context
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"github.com/zeta-chain/zetacore/pkg/chains"
+	observer "github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+// ChainRegistry is a registry of supported chains
+type ChainRegistry struct {
+	chains map[int64]Chain
+
+	// additionalChains is a list of additional static chain information to use when searching from
+	// chain IDs. It's stored in the protocol to dynamically support new chains without doing an upgrade
+	additionalChains []chains.Chain
+
+	mu sync.Mutex
+}
+
+// Chain represents chain with its parameters
+type Chain struct {
+	chainInfo      *chains.Chain
+	observerParams *observer.ChainParams
+
+	// reference to the registry it necessary for some operations
+	// like checking if the chain is EVM or not because it uses some "global" context state
```

Comment:
In this case,  all downstream services need to accept both `app zctx.AppContext` and `chain zctx.Chain`. I want to avoid this and simplify the API. 

IMO it's okay to have this connection because `AppContext.Update()` is the _only_ place to create/provision `Chain{}` and other services don't need to know the implementation behind the scenes.